scio-extra/src/main/scala/com/spotify/scio/extra/sparkey/SparkeyUri.scala

```diff
@@ -19,104 +19,138 @@ package com.spotify.scio.extra.sparkey
 
 import java.io.File
 import java.net.URI
-import java.nio.file.{Files, Paths}
-
-import com.spotify.scio.coders.Coder
 import com.spotify.scio.util.{RemoteFileUtil, ScioUtil}
+import com.spotify.scio.extra.sparkey.instances.ShardedSparkeyReader
 import com.spotify.sparkey.extra.ThreadLocalSparkeyReader
-import com.spotify.sparkey.{CompressionType, Sparkey, SparkeyReader}
-import org.apache.beam.sdk.options.PipelineOptions
+import com.spotify.sparkey.SparkeyReader
+import org.apache.beam.sdk.io.FileSystems
+import org.apache.beam.sdk.io.fs.{EmptyMatchTreatment, MatchResult, ResourceId}
 
+import java.nio.file.Path
+import scala.collection.mutable
 import scala.jdk.CollectionConverters._
 
-/**
- * Represents the base URI for a Sparkey index and log file, either on the local or a remote file
- * system. For remote file systems, `basePath` should be in the form
- * 'scheme://<bucket>/<path>/<sparkey-prefix>'. For local files, it should be in the form
- * '/<path>/<sparkey-prefix>'. Note that `basePath` must not be a folder or GCS bucket as it is a
- * base path representing two files - <sparkey-prefix>.spi and <sparkey-prefix>.spl.
- */
-trait SparkeyUri extends Serializable {
-  val basePath: String
-  def getReader: SparkeyReader
-
-  private[sparkey] def exists: Boolean
-  override def toString: String = basePath
-}
+case class InvalidShards(str: String) extends RuntimeException(str)
 
-private[sparkey] object SparkeyUri {
-  def apply(basePath: String, opts: PipelineOptions): SparkeyUri =
-    if (ScioUtil.isLocalUri(new URI(basePath))) {
-      LocalSparkeyUri(basePath)
-    } else {
-      RemoteSparkeyUri(basePath, opts)
-    }
+object SparkeyUri {
   def extensions: Seq[String] = Seq(".spi", ".spl")
-
-  implicit def coderSparkeyURI: Coder[SparkeyUri] = Coder.kryo[SparkeyUri]
+  def basePathForShard(basePath: String, shardIndex: Short, numShards: Short): String =
+    f"$basePath/part-$shardIndex%05d-of-$numShards%05d"
 }
 
-private case class LocalSparkeyUri(basePath: String) extends SparkeyUri {
-  override def getReader: SparkeyReader =
-    new ThreadLocalSparkeyReader(new File(basePath))
-  override private[sparkey] def exists: Boolean =
-    SparkeyUri.extensions.map(e => new File(basePath + e)).exists(_.exists)
-}
+case class SparkeyUri(path: String) {
+  private[sparkey] val isLocal = ScioUtil.isLocalUri(new URI(path))
+  private val isSharded = path.endsWith("*")
+  private[sparkey] val basePath =
+    if (!isSharded) path else path.split("/").dropRight(1).mkString("/")
 
-private object RemoteSparkeyUri {
-  def apply(basePath: String, options: PipelineOptions): RemoteSparkeyUri =
-    RemoteSparkeyUri(basePath, RemoteFileUtil.create(options))
-}
+  private[sparkey] def globExpression: String = s"$basePath/part-*"
```

Comment:
I think this helper is a bit risky since it should only be used when `isSharded` is `false`